modules/client/src/internal/utils.ts

```diff
@@ -824,3 +863,801 @@ export function toPluginPermissionOperationType(
+    causes.push(
+      PluginUpdateProposalInValidityCause.MISSING_PLUGIN_PREPARATION,
+    );
+  }
+  return causes;
+}
+
+/**
+ * Given a list of actions, it decodes the actions and returns the
+ * type of action
+ *
+ * @export
+ * @param {DaoAction[]} actions
+ * @return {*}  {ProposalActionTypes[]}
+ */
+export function classifyProposalActions(
+  actions: DaoAction[],
+): ProposalActionTypes[] {
+  const classifiedActions: ProposalActionTypes[] = [];
+
+  for (const action of actions) {
+    try {
+      let decodedPermission:
+        | GrantPermissionDecodedParams
+        | RevokePermissionDecodedParams;
+      const func = getFunctionFragment(action.data, UPDATE_PLUGIN_SIGNATURES);
+      switch (func.name) {
+        case "upgradeTo":
+          classifiedActions.push(ProposalActionTypes.UPGRADE_TO);
+          break;
+        case "upgradeToAndCall":
+          classifiedActions.push(ProposalActionTypes.UPGRADE_TO_AND_CALL);
+          break;
+        case "grant":
+          decodedPermission = decodeGrantAction(action.data);
+          // check the permission that is being granted
+          if (
+            decodedPermission.permission ===
+              Permissions.UPGRADE_PLUGIN_PERMISSION
+          ) {
+            classifiedActions.push(
+              ProposalActionTypes.GRANT_PLUGIN_UPDATE_PERMISSION,
+            );
+          } else if (
+            decodedPermission.permission === Permissions.ROOT_PERMISSION
+          ) {
+            classifiedActions.push(
+              ProposalActionTypes.GRANT_ROOT_PERMISSION,
+            );
+          }
+          break;
+        case "revoke":
+          decodedPermission = decodeRevokeAction(action.data);
+          // check the permission that is being granted
+          if (
+            decodedPermission.permission ===
+              Permissions.UPGRADE_PLUGIN_PERMISSION
+          ) {
+            classifiedActions.push(
+              ProposalActionTypes.REVOKE_PLUGIN_UPGRADE_PERMISSION,
+            );
+          } else if (
+            decodedPermission.permission === Permissions.ROOT_PERMISSION
+          ) {
+            classifiedActions.push(
+              ProposalActionTypes.REVOKE_ROOT_PERMISSION,
+            );
+          }
+          break;
+        case "applyUpdate":
+          classifiedActions.push(ProposalActionTypes.APPLY_UPDATE);
+          break;
+        default:
+          classifiedActions.push(ProposalActionTypes.ACTION_NOT_ALLOWED);
+          break;
+      }
+    } catch {
+      classifiedActions.push(ProposalActionTypes.UNKNOWN);
+    }
+  }
+  return classifiedActions;
+}
+
+/**
+ * Returns true if the actions are valid for a plugin update proposal with root permission
+ *
+ * @export
+ * @param {ProposalActionTypes[]} actions
+ * @return {*}  {boolean}
+ */
+export function isPluginUpdateActionWithRootPermission(
+  actions: ProposalActionTypes[],
+): boolean {
+  // get the first 5 actions
+  // if the first is a dao upgrade skip it
```

Comment:
This is not good. We should have identified the DAO upgrade separately and not pass it to the function